sentry-core/src/session.rs

```diff
@@ -184,23 +232,77 @@ impl SessionFlusher {
 
     /// Enqueues a session update for delayed sending.
     ///
-    /// When the queue is full, it will be flushed immediately.
+    /// This will aggregate session counts in request mode, for all sessions
+    /// that were not yet partially sent.
     pub fn enqueue(&self, session_update: SessionUpdate<'static>) {
         let mut queue = self.queue.lock().unwrap();
-        queue.push(session_update);
-        if queue.len() >= MAX_SESSION_ITEMS {
-            SessionFlusher::flush(queue, &self.transport);
+        if self.mode == SessionMode::Application || !session_update.init {
+            queue.individual.push(session_update);
+            if queue.individual.len() >= MAX_SESSION_ITEMS {
+                SessionFlusher::flush(queue, &self.transport);
+            }
+            return;
+        }
+
+        let aggregate = queue.aggregated.get_or_insert_with(|| AggregatedSessions {
+            buckets: HashMap::with_capacity(1),
+            attributes: session_update.attributes.clone(),
+        });
+
+        use chrono::{Duration, DurationRound};
+        let started = session_update
+            .started
+            .duration_trunc(Duration::hours(1))
+            .unwrap();
+
+        let key = AggregationKey {
+            started,
+            distinct_id: session_update.distinct_id,
+        };
+
+        let bucket = aggregate.buckets.entry(key).or_default();
+
+        match session_update.status {
+            SessionStatus::Exited => {
+                if session_update.errors > 0 {
+                    bucket.errored += 1;
+                } else {
+                    bucket.exited += 1;
+                }
+            }
+            SessionStatus::Crashed => {
+                bucket.crashed += 1;
+            }
+            SessionStatus::Abnormal => {
+                bucket.abnormal += 1;
+            }
+            SessionStatus::Ok => {
+                unreachable!("only closed sessions will be enqueued");
+            }
         }
     }
 
     /// Flushes the queue to the transport.
     ///
     /// This is a static method as it will be called from both the background
     /// thread and the main thread on drop.
-    fn flush(mut queue_lock: MutexGuard<Vec<SessionUpdate<'static>>>, transport: &TransportArc) {
-        let queue: Vec<_> = std::mem::take(queue_lock.as_mut());
+    fn flush(mut queue_lock: MutexGuard<SessionQueue>, transport: &TransportArc) {
+        let (queue, aggregate) = (
+            std::mem::take(&mut queue_lock.individual),
+            queue_lock.aggregated.take(),
```

Comment:
Since you have a mutex on the queue, can you assign this in two regular statements?